raft.go

```diff
@@ -652,6 +702,75 @@ func (r *Raft) verifyLeader(v *verifyFuture) {
 	}
 }
 
+// leadershipTransfer is doing the heavy lifting for the leadership transfer.
+func (r *Raft) leadershipTransfer(future *leadershipTransferFuture) {
+
+	// Step 1: set this field which stops this leader from responding to any client requests.
+	r.leaderState.leadershipTransferInProgress = true
+
+	// Step 2: make sure the target server is up to date
+	s, ok := r.leaderState.replState[future.ID]
+	if !ok {
+		err := fmt.Errorf("cannot find replication state for %v, aborting leadership transfer", future)
+		r.logger.Printf("[DEBUG] raft: %v", err)
+		future.respond(err)
+		return
+	}
+
+	for s.nextIndex <= r.getLastIndex() {
+		r.logger.Printf("[DEBUG] raft: waiting for replication to catch up before sending TimeoutNow")
+		err := &deferError{}
+		err.init()
+		s.triggerDeferErrorCh <- err
+		if err.Error() != nil {
+			r.logger.Printf("[DEBUG] raft: %v", err.Error())
+			future.respond(err.Error())
+			return
+		}
+	}
+
+	// Step 3: reset leader lease. Since a new leader might be choosen in a bit,
+	// we don't want this leader to have the right to answer read request based on
+	// this lease.
+	r.leaderState.lease = time.After(0)
+
+	// Track leadership status with the notify channel.
+	verify := &verifyFuture{}
+	verify.init()
+	r.leaderState.notify[verify] = struct{}{}
+
+	// When leadership is lost, the caller is being notified via the future. When
+	// leadership is not lost after an election timeout, the current leader
+	// resumes operation and continues as if nothing has happened. We are setting
+	// everything up here, although we yet have to send TimeoutNowRequest. This is
+	// being done so that we don't miss leadership loss.
+	go func() {
+		select {
+		case err := <-verify.errCh:
+			r.logger.Printf("[DEBUG] raft: resetting leadership transfer: %s", err)
+			future.respond(nil)
+			r.leaderState.leadershipTransferInProgress = false
```

Comment:
I put this into a defer: https://github.com/hashicorp/raft/blob/c2eb642215d568b9925f86781a9b629c5fd1457a/raft.go#L755-L759

But since it is still being read from the mainloop, I also put it into an atomic int.